lib/curl/logger.ex

```diff
@@ -1,220 +1,51 @@
-        {:ok, state}
-
-      is_nil(ref) ->
-        {:ok, log_event(level, msg, ts, md, state)}
-
-      buffer_size < max_buffer ->
-        {:ok, buffer_event(level, msg, ts, md, state)}
-
-      buffer_size === max_buffer ->
-        state = buffer_event(level, msg, ts, md, state)
-        {:ok, await_io(state)}
-    end
-  end
-
-  def handle_event(:flush, state) do
-    {:ok, flush(state)}
-  end
-
-  def handle_event(_, state) do
-    {:ok, state}
-  end
-
-  def handle_info({:io_reply, ref, msg}, %{ref: ref} = state) do
-    {:ok, handle_io_reply(msg, state)}
-  end
-
-  def handle_info({:DOWN, ref, _, pid, reason}, %{ref: ref}) do
-    raise "device #{inspect(pid)} exited: " <> Exception.format_exit(reason)
-  end
-
-  def handle_info(_, state) do
-    {:ok, state}
-  end
-
-  def code_change(_old_vsn, state, _extra) do
-    {:ok, state}
-  end
-
-  def terminate(_reason, _state) do
-    :ok
-  end
-
-  ## Helpers
-
-  defp meet_level?(_lvl, nil), do: true
-
-  defp meet_level?(lvl, min) do
-    Logger.compare_levels(lvl, min) != :lt
-  end
-
-  defp configure(options, state) do
-    config = Keyword.merge(Application.get_env(:logger, __MODULE__), options)
-    Application.put_env(:logger, __MODULE__, config)
-    init(config, state)
-  end
-
-  defp init(config, state) do
-    level = Keyword.get(config, :level)
-    service = Keyword.get(config, :service)
-    device = Keyword.get(config, :device, :user)
-    metadata = Keyword.get(config, :metadata, []) |> configure_metadata()
-    max_buffer = Keyword.get(config, :max_buffer, 32)
-
-    %{
-      state
-      | metadata: metadata,
-        level: level,
-        service: service,
-        device: device,
-        max_buffer: max_buffer
-    }
-  end
-
-  defp configure_metadata(:all), do: :all
-  defp configure_metadata(metadata), do: Enum.reverse(metadata)
-
-  defp log_event(level, msg, ts, md, %{device: device} = state) do
-    output = format_event(level, msg, ts, md, state)
-    %{state | ref: async_io(device, output), output: output}
-  end
-
-  defp buffer_event(level, msg, ts, md, state) do
-    %{buffer: buffer, buffer_size: buffer_size} = state
-    buffer = [buffer | format_event(level, msg, ts, md, state)]
-    %{state | buffer: buffer, buffer_size: buffer_size + 1}
-  end
-
-  defp async_io(name, output) when is_atom(name) do
-    case Process.whereis(name) do
-      device when is_pid(device) ->
-        async_io(device, output)
-
-      nil ->
-        raise "no device registered with the name #{inspect(name)}"
+  def format(level, message, timestamp, metadata) do
+    try do
+      message
+      |> to_map
+      |> add_timestamp(timestamp)
+      |> add_metadata(metadata)
+      |> add_level(level)
+      |> add_service
+      |> to_json
+    rescue
+      _ ->
+        case Poison.encode(%{message: "could not format: #{inspect message}}", level: level, metadata: inspect(metadata)}) do
+          {:ok, message} -> message
+          _ -> ""
```

Comment:
If this case happens, we probably want some fallback to see that it is happening and what's causing it, we wouldn't want to fly blind.

Reply:
@ammbot I think we should leave something informative here rather than empty string. perhaps make use of `inspect`

Reply:
OK. but we definitely shouldn't have empty string. perhaps just change the string to "unexpected outcome" then at least if we see it in logs we can get to the source